### PR TITLE
Row write support

### LINKED ISSUE #343: [FEATURE] Support Row write
## Description
Implement writing Spark sql Structs as Vertica rows

### PR DESCRIPTION
**Pull request by Aryex**
### Summary
Added row write support 

### Description

SchemaTools can now parse Spark struct to Vertica Rows
Added support for Arrays[Rows]
Examples to be added in another PR

### Related Issue

Closes #343

### Additional Reviewers

@alexey-temnikov 
@alexr-bq 
@jeremyp-bq 
@jonathanl-bq

### CODE REVIEW

**Review by alexr-bq — APPROVED** [review]

File: connector/src/main/scala/com/vertica/spark/util/error/ErrorHandling.scala [hunk 1]
```diff
@@ -472,6 +472,11 @@ case class ParamsNotSupported(operation: String) extends JdbcError {
   */
 trait SchemaError extends ConnectorError
 
+case class StructFieldsError(error: ConnectorError) extends SchemaError {
+  def getFullContext: String = s"${this.error.getClass.getSimpleName}: ${error.getFullContext}"
+  override def getUserMessage: String = s"${this.error.getClass.getSimpleName}: ${error.getFullContext}"
```
Comment (jonathanl-bq) [inline connector/src/main/scala/com/vertica/spark/util/error/ErrorHandling.scala:477]: Did this need to be overridden? I think the default is to just call getFullContext anyway, which this appears to basically be doing.

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala [hunk 2]
```diff
@@ -399,13 +399,25 @@ class SchemaTools extends SchemaToolsInterface {
...
-           org.apache.spark.sql.types.StructType(_) => Right("VARBINARY(" + longlength + ")")
+      case org.apache.spark.sql.types.MapType(_,_,_) => Right("VARBINARY(" + longlength + ")")
+      case org.apache.spark.sql.types.StructType(fields) => sparkStructToVerticaRow(fields, strlen, arrayLength)
       case org.apache.spark.sql.types.ArrayType(sparkType,_) => sparkArrayToVerticaArray(sparkType, strlen, arrayLength)
       case _ => this.sparkPrimitiveToVerticaPrimitive(sparkType, strlen)
     }
   }
 
+  private def sparkStructToVerticaRow(fields: Array[StructField], strlen: Long, arrayLength: Long): SchemaResult[String] = {
+    makeTableColumnDefs(StructType(fields), strlen, null, arrayLength) match {
```
Comment (jonathanl-bq) [inline connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:410]: Why are we passing in null here?

**Review by jonathanl-bq — CHANGES_REQUESTED** [review]

File: connector/src/main/scala/com/vertica/spark/util/error/ErrorHandling.scala (hunk 1 above)
Comment (Aryex) [reply connector/src/main/scala/com/vertica/spark/util/error/ErrorHandling.scala:477]: Good call. I was copying from other errors so I didn't realized it default to getFullContext.

File: connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala (hunk 2 above)
Comment (Aryex) [reply connector/src/main/scala/com/vertica/spark/util/schema/SchemaTools.scala:410]: Fixed. The argument was for a jdbc layer but it is not being used at all. It was added from PR #336 so must've been an oversight on my end.

**Review by jonathanl-bq — APPROVED** [review]

**Pull request merged**